symfit/core/support.py

```diff
@@ -50,45 +53,72 @@ def deprecated_func(*args, **kwargs):
             return func(*args, **kwargs)
         return deprecated_func
 
-def seperate_symbols(func):
+def seperate_symbols(func, separate_indices=False):
     """
     Seperate the symbols in symbolic function func. Return them in alphabetical
     order.
 
     :param func: scipy symbolic function.
-    :return: (vars, params), a tuple of all variables and parameters, each 
-        sorted in alphabetical order.
+    :param separate_indices: If True, Indexed objects are separated into their
+        IndexedBase and Idx, and apart from variables and parameter, also a list
+        of all indices is returned. If Flase, Indexed object are seen as
+        fundamental.
+    :return: (vars, params) if `separate_indices` is False, or
+        (vars, params, indices) when True.
+        each of these items is an alphabetically ordered list.
     :raises TypeError: only symfit Variable and Parameter are allowed, not sympy
         Symbols.
     """
     params = []
     vars = []
+    indices = []
     for symbol in func.free_symbols:
         if isinstance(symbol, Parameter):
             params.append(symbol)
-        elif isinstance(symbol, Idx):
-            # Idx objects are not seen as parameters or vars.
-            pass
+        elif isinstance(symbol, IndexedParameter):
+            if separate_indices:
+                params.append(symbol.base)
+                indices.extend(symbol.indices)
+            else:
+                params.append(symbol)
+        elif isinstance(symbol, IndexedParameterBase):
+            params.append(symbol)
+        elif isinstance(symbol, IndexedVariable):
+            if separate_indices:
+                vars.append(symbol.base)
+                indices.extend(symbol.indices)
+            else:
+                vars.append(symbol)
+        elif isinstance(symbol, IndexedVariableBase):
+            vars.append(symbol)
```

Comment:
You can clean this if/elif tree by making a dict of {classes: lists}, and sorting based on that. At the end you combine some (possibly also based on lists/sets of classes).
```python3
outcomes = {Parameter: [], IndexedParameter: [], IndexedParameterBase: [], ...}
for symbol in func.free_symbols:
    for class_, items in outcomes.items():
        if isinstance(symbol, class_):
            items.append(symbol)
```
etc.